File: pixels-core/src/main/java/io/pixelsdb/pixels/core/vector/LongColumnVector.java
/*
 * Copyright 2017-2019 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
package io.pixelsdb.pixels.core.vector;

import io.pixelsdb.pixels.core.utils.Bitmap;

import java.util.Arrays;

import static com.google.common.base.Preconditions.checkArgument;
import static java.util.Objects.requireNonNull;

/**
 * LongColumnVector derived from org.apache.hadoop.hive.ql.exec.vector
 * <p>
 * This class represents a nullable int column vector.
 * This class will be used for operations on all integer types (tinyint, smallint, int, bigint)
 * and as such will use a 64-bit long value to hold the biggest possible value.
 * During copy-in/copy-out, smaller int types will be converted as needed. This will
 * reduce the amount of code that needs to be generated and also will run fast since the
 * machine operates with 64-bit words.
 * <p>
 * The vector[] field is public by design for high-performance access in the inner
 * loop of query execution.
 */
public class LongColumnVector extends ColumnVector
{
    public long[] vector;

    /**
     * Use this constructor by default. All column vectors
     * should normally be the default size.
     */
    public LongColumnVector()
    {
        this(VectorizedRowBatch.DEFAULT_SIZE);
    }

    /**
     * Don't use this except for testing purposes.
     *
     * @param len the number of rows
     */
    public LongColumnVector(int len)
    {
        super(len);
        vector = new long[len];
        memoryUsage += Long.BYTES * len;
    }

    @Override
    public void add(String value)
    {
        switch (value.toLowerCase())
        {
            case "true":
                add(1);
                break;
            case "false":
                add(0);
                break;
            default:
                add(Long.parseLong(value));
                break;
        }
    }

    @Override
    public void add(boolean value)
    {
        add(value ? 1 : 0);
    }

    @Override
    public void add(long v)
    {
        if (writeIndex >= getLength())
        {
            ensureSize(writeIndex * 2, true);
        }
        int index = writeIndex++;
        vector[index] = v;
        isNull[index] = false;
    }

    @Override
    public void add(int v)
    {
        if (writeIndex >= getLength())
        {
            ensureSize(writeIndex * 2, true);
        }
        int index = writeIndex++;
        vector[index] = v;
        isNull[index] = false;
    }

    @Override
    public int[] accumulateHashCode(int[] hashCode)
    {
        requireNonNull(hashCode, "hashCode is null");
        checkArgument(hashCode.length > 0 && hashCode.length <= this.length, "",
                "the length of hashCode is not in the range [1, length]");
        for (int i = 0; i < hashCode.length; ++i)
        {
            if (this.isNull[i])
            {
                continue;
            }
            hashCode[i] = 31 * hashCode[i] + (int)(this.vector[i] ^ (this.vector[i] >>> 16));
        }
        return hashCode;
    }

    @Override
    public boolean elementEquals(int index, int otherIndex, ColumnVector other)
    {
        LongColumnVector otherVector = (LongColumnVector) other;
        if (!this.isNull[index] && !otherVector.isNull[otherIndex])
        {
            return this.vector[index] == otherVector.vector[otherIndex];
        }
        return false;
    }

    @Override
    public int compareElement(int index, int otherIndex, ColumnVector other)
    {
        LongColumnVector otherVector = (LongColumnVector) other;
        if (!this.isNull[index] && !otherVector.isNull[otherIndex])
        {
            return this.vector[index] < otherVector.vector[otherIndex] ? -1 :
                    (this.vector[index] == otherVector.vector[otherIndex] ? 0 : 1);
        }
        return this.isNull[index] ? -1 : 1;
    }

    // Fill the column vector with the provided value
    public void fill(long value)
    {
        noNulls = true;
        isRepeating = true;
        vector[0] = value;
    }

    // Simplify vector by brute-force flattening noNulls and isRepeating
    // This can be used to reduce combinatorial explosion of code paths in VectorExpressions
    // with many arguments.
    public void flatten(boolean selectedInUse, int[] sel, int size)
    {
        flattenPush();
        if (isRepeating)
        {
            isRepeating = false;
            long repeatVal = vector[0];
            if (selectedInUse)
            {
                for (int j = 0; j < size; j++)
                {
                    int i = sel[j];
                    vector[i] = repeatVal;
                }
            }
            else
            {
                Arrays.fill(vector, 0, size, repeatVal);
            }
            writeIndex = size;
            flattenRepeatingNulls(selectedInUse, sel, size);
        }
        flattenNoNulls(selectedInUse, sel, size);
    }

    @Override
    public void addElement(int inputIndex, ColumnVector inputVector)
    {
        int index = writeIndex++;
        if (inputVector.noNulls || !inputVector.isNull[inputIndex])
        {
            isNull[index] = false;
            vector[index] = ((LongColumnVector) inputVector).vector[inputIndex];
        }
        else
        {
            isNull[index] = true;
            noNulls = false;
        }
    }

    @Override
    public void addSelected(int[] selected, int offset, int length, ColumnVector src)
    {
        // isRepeating should be false and src should be an instance of LongColumnVector.
        // However, we do not check these for performance considerations.
        LongColumnVector source = (LongColumnVector) src;

        for (int i = offset; i < offset + length; i++)
        {
            int srcIndex = selected[i], thisIndex = writeIndex++;
            if (source.isNull[srcIndex])
            {
                this.isNull[thisIndex] = true;
                this.noNulls = false;
            }
            else
            {
                this.vector[thisIndex] = source.vector[srcIndex];
                this.isNull[thisIndex] = false;
            }
        }
    }

    @Override
    public void duplicate(ColumnVector inputVector)
    {
        if (inputVector instanceof LongColumnVector)
        {
            LongColumnVector srcVector = (LongColumnVector) inputVector;
            this.vector = srcVector.vector;
            this.isNull = srcVector.isNull;
            this.writeIndex = srcVector.writeIndex;
            this.noNulls = srcVector.noNulls;
            this.isRepeating = srcVector.isRepeating;
        }
    }

    @Override
    protected void applyFilter(Bitmap filter, int before)
    {
        checkArgument(!isRepeating,
                "column vector is repeating, flatten before applying filter");
        checkArgument(before > 0 && before <= length,
                "before index is not in the range [1, length]");
        boolean noNulls = true;
        int j = 0;
        for (int i = filter.nextSetBit(0);
             i >= 0 && i < before; i = filter.nextSetBit(i+1), j++)
        {
            if (i > j)
            {
                this.vector[j] = this.vector[i];
                this.isNull[j] = this.isNull[i];
            }
            if (this.isNull[j])
            {
                noNulls = false;
            }
            /*
             * The number of rows in a row batch is impossible to reach Integer.MAX_VALUE.
             * Therefore, we do not check overflow here.
             */
        }
        this.noNulls = noNulls;
    }

    @Override
    public void stringifyValue(StringBuilder buffer, int row)
    {
        if (isRepeating)
        {
            row = 0;
        }
        if (noNulls || !isNull[row])
        {
            buffer.append(vector[row]);
        }
        else
        {
            buffer.append("null");
        }
    }

    @Override
    public void ensureSize(int size, boolean preserveData)
    {
        super.ensureSize(size, preserveData);
        if (size > vector.length)
        {
            long[] oldArray = vector;
            vector = new long[size];
            memoryUsage += Long.BYTES * size;
            length = size;
            if (preserveData)
            {
                if (isRepeating)
                {
                    vector[0] = oldArray[0];
                }
                else
                {
                    System.arraycopy(oldArray, 0, vector, 0, oldArray.length);
                }
            }
        }
    }

    @Override
    public LongColumnVector clone()
    {
        try
        {
            LongColumnVector cloned = (LongColumnVector) super.clone();
            cloned.vector = this.vector == null ? null : this.vector.clone();
            return cloned;
        }
        catch (CloneNotSupportedException e)
        {
            throw new AssertionError();
        }
    }

    @Override
    public void close()
    {
        super.close();
        this.vector = null;
    }
}

File: cpp/pixels-core/lib/vector/LongColumnVector.cpp
/*
 * Copyright 2023 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */

/*
 * @author liyu
 * @create 2023-03-17
 */
#include "vector/LongColumnVector.h"
#include <algorithm>

LongColumnVector::LongColumnVector(uint64_t len, bool encoding, bool isLong)
        : ColumnVector (len, encoding)
{
    posix_memalign (reinterpret_cast<void **>(&longVector), 32,
                    len * sizeof (int64_t));

    memoryUsage += (long) sizeof (long) * len;
}

void LongColumnVector::close()
{
    if (!closed)
    {
        ColumnVector::close ();
        if (encoding && longVector != nullptr)
        {
            free (longVector);
        }
        longVector = nullptr;
    }
}

void LongColumnVector::print(int rowCount)
{
    throw InvalidArgumentException ("not support print longcolumnvector.");
    //    for(int i = 0; i < rowCount; i++) {
    //        std::cout<<longVector[i]<<std::endl;
    //		std::cout<<intVector[i]<<std::endl;
    //    }
}

LongColumnVector::~LongColumnVector()
{
    if (!closed)
    {
        LongColumnVector::close ();
    }
}

void *LongColumnVector::current()
{

    if (longVector == nullptr)
    {
        return nullptr;
    } else
    {
        return longVector + readIndex;
    }
}

void LongColumnVector::add(std::string &value)
{
    std::transform (value.begin (), value.end (), value.begin (), ::tolower);
    if (value == "true")
    {
        add (1);
    } else if (value == "false")
    {
        add (0);
    } else
    {
        add (std::stol (value));
    }
}

void LongColumnVector::add(bool value) { add (value ? 1 : 0); }

void LongColumnVector::add(int64_t value)
{
    if (writeIndex >= length)
    {
        ensureSize (writeIndex * 2, true);
    }
    int index = writeIndex++;

    longVector[index] = value;

    isNull[index] = false;
}

void LongColumnVector::add(int value)
{
    if (writeIndex >= length)
    {
        ensureSize (writeIndex * 2, true);
    }
    int index = writeIndex++;
    longVector[index] = value;

    isNull[index] = false;
}

void LongColumnVector::ensureSize(uint64_t size, bool preserveData)
{
    ColumnVector::ensureSize (size, preserveData);
    if (length < size)
    {

        long *oldVector = longVector;
        posix_memalign (reinterpret_cast<void **>(&longVector), 32,
                        size * sizeof (int64_t));
        if (preserveData)
        {
            std::copy (oldVector, oldVector + length, longVector);
        }
        delete[] oldVector;
        memoryUsage += (long) sizeof (long) * (size - length);
        resize (size);
    }
}
